src/docker/dockerUtils.ts

```diff
@@ -0,0 +1,110 @@
+import * as fs from "fs";
+import * as path from "path";
+import * as vscode from "vscode";
+import * as url from "url";
+
+import { kubeChannel } from "../kubeChannel";
+import { Kubectl } from "../kubectl";
+import { getCurrentClusterConfig } from "../kubectlUtils";
+import { shell, ShellResult } from "../shell";
+import { DockerfileParser } from "./dockerfileParser";
+
+export enum DockerClient {
+    docker = "docker",
+    dockerCompose = "docker-compose"
+}
+
+/**
+ * Build the docker image first. If imagePrefix is not empty, push the image to remote docker hub, too.
+ * 
+ * @param dockerClient the possible dockerClient： docker or docker-compose.
+ * @param shellOpts any option available to Node.js's child_process.exec().
+ * @param imagePrefix the image prefix for docker images (e.g. 'docker.io/brendanburns').
+ * @return the image name.
+ */
+export async function buildAndPushDockerImage(dockerClient: DockerClient, shellOpts: any, imagePrefix?: string): Promise<string> {
+    const cwd = shellOpts.cwd || vscode.workspace.rootPath;
+    const image = await getDefaultImageName(cwd, imagePrefix);
+    await buildDockerImage(dockerClient, image, shellOpts);
+    if (imagePrefix) {
+        await pushDockerImage(dockerClient, image, shellOpts);
+    }
+    return image;
+}
+
+function sanitiseTag(name : string) {
+    // Name components may contain lowercase letters, digits and separators.
+    // A separator is defined as a period, one or two underscores, or one or
+    // more dashes. A name component may not start or end with a separator.
+    // https://docs.docker.com/engine/reference/commandline/tag/#extended-description
+
+    return name.toLowerCase().replace(/[^a-z0-9._-]/g, '-');
+}
+
+async function getDefaultImageName(cwd: string, imagePrefix?: string): Promise<string> {
+    const name = sanitiseTag(path.basename(cwd));
+    const version = await findVersion(cwd);
+    let image = `${name}:${version}`;
+    if (imagePrefix) {
+        image = imagePrefix + "/" + image;
+    }
+    return image;
+}
+
+async function findVersion(cwd: string): Promise<string> {
+    const shellOpts = Object.assign({ }, shell.execOpts(), { cwd });
+    const shellResult = await shell.execCore('git describe --always --dirty', shellOpts);
+    return shellResult.code !== 0 ? "latest" : shellResult.stdout.trim();

```

Comment:
Reverse this condition to avoid the reader having to go, "if it's not zero then ..., but if it's _not_ not zero..."